core/src/main/java/com/linecorp/armeria/common/MediaTypeSet.java

```diff
@@ -0,0 +1,345 @@
+     * specified string.
+     *
+     * @param acceptHeaders the values of the {@code "accept"} header, as defined in
+     *        <a href="https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html">the section 14.1, RFC2616</a>
+     * @return the most preferred {@link MediaType} that matches one of the specified media ranges.
+     *         {@link Optional#empty()} if there are no matches or {@code acceptHeaders} does not contain
+     *         any valid ranges.
+     */
+    public Optional<MediaType> matchHeaders(Iterable<? extends CharSequence> acceptHeaders) {
+        requireNonNull(acceptHeaders, "acceptHeaders");
+
+        final List<MediaType> ranges = new ArrayList<>(4);
+        for (CharSequence acceptHeader : acceptHeaders) {
+            addRanges(ranges, acceptHeader);
+        }
+        return match(ranges);
+    }
+
+    /**
+     * Finds the {@link MediaType} in this {@link List} that matches one of the media ranges specified in the
+     * specified string.
+     *
+     * @param acceptHeaders the values of the {@code "accept"} header, as defined in
+     *        <a href="https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html">the section 14.1, RFC2616</a>
+     * @return the most preferred {@link MediaType} that matches one of the specified media ranges.
+     *         {@link Optional#empty()} if there are no matches or {@code acceptHeaders} does not contain
+     *         any valid ranges.
+     */
+    public Optional<MediaType> matchHeaders(CharSequence... acceptHeaders) {
+        requireNonNull(acceptHeaders, "acceptHeaders");
+
+        final List<MediaType> ranges = new ArrayList<>(4);
+        for (CharSequence acceptHeader : acceptHeaders) {
+            addRanges(ranges, acceptHeader);
+        }
+        return match(ranges);
+    }
+
+    private Optional<MediaType> match(List<MediaType> ranges) {
+        return ranges.isEmpty() ? Optional.empty()
+                                : match(ranges.toArray(new MediaType[ranges.size()]));
+    }
+
+    /**
+     * Finds the {@link MediaType} in this {@link List} that matches one of the specified media ranges.
+     *
+     * @return the most preferred {@link MediaType} that matches one of the specified media ranges.
+     *         {@link Optional#empty()} if there are no matches or {@code ranges} does not contain
+     *         any valid ranges.
+     */
+    public Optional<MediaType> match(Iterable<MediaType> ranges) {
+        return match(Iterables.toArray(requireNonNull(ranges, "ranges"), MediaType.class));
+    }
+
+    /**
+     * Finds the {@link MediaType} in this {@link List} that matches one of the specified media ranges.
+     *
+     * @return the most preferred {@link MediaType} that matches one of the specified media ranges.
+     *         {@link Optional#empty()} if there are no matches or {@code ranges} does not contain
+     *         any valid ranges.
+     */
+    public Optional<MediaType> match(MediaType... ranges) {
```

Comment:
Since the common case seems to be a single range, consider implementing the logic against a List instead of array - a caller can pass in a singleton ImmutableList which is just a simple struct so can be optimized better.